library/core/src/iter/traits/iterator.rs

```diff
@@ -3613,29 +3595,26 @@ pub trait Iterator {
     /// assert!(xs.iter().eq_by(&ys, |&x, &y| x * x == y));
     /// ```
     #[unstable(feature = "iter_order_by", issue = "64295")]
-    fn eq_by<I, F>(mut self, other: I, mut eq: F) -> bool
+    fn eq_by<I, F>(self, other: I, eq: F) -> bool
```

Comment:
Probably not this PR, but it might be interesting to experiment with a `size_hint` check for this one.

Something like
```rust
pub fn definitely_not_equal(a: &impl Iterator, b: &impl Iterator) -> bool {
    match (a.size_hint(), b.size_hint()) {
        ((a_lo, _), (_, Some(b_hi))) => b_hi < a_lo,
        ((_, Some(a_hi)), (b_lo, _)) => a_hi < b_lo,
        _ => false,
    }
}
```

Since that could instantly catch a bunch of cases where things have good size_hints in practice, while also optimizing away completely for things that always return `(0, None)` since that'd obviously always be `false` overall.

Reply:
For sure! I almost included that in this PR, in fact. It's a behavior change, but hopefully one we feel like we're permitted to make.